### ISSUE TITLE
Missing required module '_FoundationCollections' when building with the Static Linux SDK

### ISSUE THREAD

**Issue by barnard-b** [apple/swift-collections#420]
When building a Swift package using the September 11, 2024 Swift 6 toolchain and matching Static Linux SDK, compilation fails with `error: missing required module '_FoundationCollections'`

```bash
% xcrun --toolchain swift swift build --swift-sdk aarch64-swift-linux-musl   
Building for debugging...
error: emit-module command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: missing required module '_FoundationCollections'
<unknown>:0: error: missing required module '_FoundationCollections'
```

## System info
**OS:** macOS 14.6.1
**Xcode:** 16.0 public release, 16.0 release candidate, 15.4
**Swift**: Apple Swift version 6.0-dev (LLVM fc1036139c5e562, Swift 5bbf725fd958d34) Target: arm64-apple-macosx14.0
**Static Linux SDK** swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-11-a_static-linux-0.0.1

## Package
**Package.swift**
```swift
// swift-tools-version: 6.0

import PackageDescription

let package = Package(
    name: "FoundationTest",
    targets: [
        .executableTarget(name: "FoundationTest")
    ]
)
```

**Sources/FoundationTest/FoundationTest.swift**
```swift
import Foundation

@main
struct App {
    static func main() async throws {
        print("Hello World!")
    }
}
```

**Comment by finagolfin** [comment]
@al45tair, others are reporting more errors like `<unknown>:0: warning: libc not found for 'x86_64-swift-linux-musl'; C stdlib may be unavailable`, apple/swift-nio#2886, and `error: unableToFind(tool: "swift-autolink-extract")`, maybe [because they're wrongly using the Xcode toolchain](https://forums.swift.org/t/trouble-cross-compiling-with-swift-6-0/74658/4)?

You may also want to [update the doc](https://www.swift.org/documentation/articles/static-linux-getting-started.html) to use the 6.0 release SDK and add a doc step so they can check and make sure they're not using the Xcode toolchain inadvertently.

**Comment by al45tair** [comment]
The warning is a known issue with the modulewrap program (rdar://115918181); it's harmless, but annoying. Somewhat irritating that it hasn't been fixed yet, mind.

The missing `swift-autolink-extract` is almost certainly, as you say, because they're using the Xcode toolchain. Updating the documentation isn't a bad idea, but let's make sure everything works first.

**Comment by doozMen** [comment]
The thing I had was that I overlooked the line on https://www.swift.org/documentation/articles/static-linux-getting-started.html 
> You cannot use the toolchain provided with Xcode to build programs using the SDK.

So you have to follow the instructions to build using the toolchain downloaded and not build xcode. That is indeed what solved the `swift-autolink-extract`.

But importing Foundation is an issue. It does not work as documented like stated in this post https://forums.swift.org/t/using-swift-foundation-with-linux-static-sdk/74642

Also doing `import FoundationEssentials` with the 6.0 toolchain from cli or from xcode when that toolchain is selected gives error No such module 'FoundationEssentials', on macOS 15.0 and on linux. 

So also the docs on apple/swift-foundation that state it should be included are wrong. Any help why foundation is missing?

**Comment by al45tair** [comment]
> So also the docs on apple/swift-foundation that state it should be included are wrong. Any help why foundation is missing?

Foundation itself is not missing, but `_FoundationCollections` _is_, which breaks things (`_FoundationCollections` is a dependency of `FoundationEssentials`, FWIW). This is fixed by https://github.com/apple/swift-collections/pull/421, but that will need to work its way through to a release in order to cure this problem.

**Comment by finagolfin** [comment]
> Also doing import FoundationEssentials with the 6.0 toolchain from cli or from xcode when that toolchain is selected gives error No such module 'FoundationEssentials', on macOS 15.0 and on linux.

Hmm, that is new. Can you try something like the following two commands on your linux toolchain and report the results?
```
> ./swift-6.0-RELEASE-fedora39/usr/bin/swift -version
> find ./swift-6.0-RELEASE-fedora39/usr/lib/swift/linux -name "*.swiftdoc"
```

**Comment by shahmishal** [comment]
Swift 6.0 Nightly should resolve this  issue with temporary workaround in the build script.
`error: missing required module '_FoundationCollections'`

We would like to hear feedback if this resolves the issue for you. 

**Toolchain**: https://download.swift.org/swift-6.0-branch/xcode/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a-osx.pkg
**Static SDK**: [https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-[…]PSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz](https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz) (b34ae1758ba66f2cd0d8c7e93b504df199905ddd69dabc7e6b686542b2c43547)

Note: We should not close this issue until the proper fix is merged into swift-collections repo.

**Comment by doozMen** [comment]
> > Also doing import FoundationEssentials with the 6.0 toolchain from cli or from xcode when that toolchain is selected gives error No such module 'FoundationEssentials', on macOS 15.0 and on linux.
> 
> Hmm, that is new. Can you try something like the following two commands on your linux toolchain and report the results?
> 
> ```
> > ./swift-6.0-RELEASE-fedora39/usr/bin/swift -version
> > find ./swift-6.0-RELEASE-fedora39/usr/lib/swift/linux -name "*.swiftdoc"
> ```

Not sure I understand the find, this does not fin anything but the swift --version gives this 
swift --version
Apple Swift version 6.0 (swift-6.0-RELEASE)
Target: arm64-apple-macosx15.0

**Comment by doozMen** [comment]
@shahmishal will try with the de

> Swift 6.0 Nightly should resolve this issue with temporary workaround in the build script. `error: missing required module '_FoundationCollections'`
> 
> We would like to hear feedback if this resolves the issue for you.
> 
> **Toolchain**: https://download.swift.org/swift-6.0-branch/xcode/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a-osx.pkg **Static SDK**: [https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-[…]PSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz](https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz) (b34ae1758ba66f2cd0d8c7e93b504df199905ddd69dabc7e6b686542b2c43547)
> 
> Note: We should not close this issue until the proper fix is merged into swift-collections repo.

will try using those now and report back ...

**Comment by doozMen** [comment]
I think I'm mixing 2 issues that should be separated.

1. the static linux build seams to not work when Foundation is imported
2. the normal build of any project that imports Foundation does not build when swift 6 release of  `swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a-osx.pkg` is used

This is for a very simple executable project that I mention in the ticket in Foundation. That has been closed since it seams to be a darwin related issue

https://github.com/apple/swift-foundation/issues/920

the reproducable code to try is 

```
export TOOLCHAINS=$(plutil -extract CFBundleIdentifier raw /Library/Developer/Toolchains/swift-6.0-RELEASE.xctoolchain/Info.plist)
mkdir swift-tool && cd swift-tool
swift package init --type executable
echo `import Foundation` >> Sources/main.swift
swift build
```

This both with  with release and the swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a.xctoolchain

**Comment by shahmishal** [comment]
Correct there are two issues: 
1. static sdk running into `error: missing required module '_FoundationCollections'`
    * Fixed in swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a nightly tag. 
    * To verify this fix, you will need to use Xcode 16 beta 4 or Xcode 15.x otherwise you will run into issue number 2. 
2. Unable to build with Xcode 16 + Swift.org toolchain due to Darwin module 
    * Fixed with https://github.com/swiftlang/swift/pull/76552, however nightly toolchain has not picked up this fix yet.

**Comment by finagolfin** [comment]
> Not sure I understand the find, this does not fin anything

Nobody else has reported the `No such module 'FoundationEssentials'` error that you saw, so I'm trying to figure out if that's a third problem, particularly on linux. Could you detail how you see that when building for linux, ie what code you're compiling and what toolchain and SDK you're using?

I have seen something similar when cross-compiling the 6.0 corelibs for Android and trying to use them, so I'm trying to figure out if you're hitting that on certain linux configurations also.

**Comment by barnard-b** [comment]
> Swift 6.0 Nightly should resolve this issue with temporary workaround in the build script. `error: missing required module '_FoundationCollections'`
> 
> We would like to hear feedback if this resolves the issue for you.
> 
> **Toolchain**: https://download.swift.org/swift-6.0-branch/xcode/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a-osx.pkg **Static SDK**: [https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-[…]PSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz](https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz) (b34ae1758ba66f2cd0d8c7e93b504df199905ddd69dabc7e6b686542b2c43547)

@shahmishal When I attempt to install the new Static SDK it says the checksum does not match. I get the same different hash when downloading the file from https://www.swift.org/install/macos/ in Safari and running `shasum`

```
% shasum -a 256 ./swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz
83a88650cd0675552ce2cf8159c31966fde73418f49493c1644073fffe8be9f4
```

```
% xcrun --toolchain swift swift sdk install https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz --checksum "b34ae1758ba66f2cd0d8c7e93b504df199905ddd69dabc7e6b686542b2c43547"
Downloading a Swift SDK bundle archive from `https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz`...
                                                                        Downloading
100% [=======================================================================================================================================]
Downloading swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz

Swift SDK bundle archive successfully downloaded from `https://download.swift.org/swift-6.0-branch/static-sdk/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1.artifactbundle.tar.gz`.
Verifying if checksum of the downloaded archive is valid...
Error: Computed archive checksum `83a88650cd0675552ce2cf8159c31966fde73418f49493c1644073fffe8be9f4` does not match the provided checksum `b34ae1758ba66f2cd0d8c7e93b504df199905ddd69dabc7e6b686542b2c43547`.
```

**Comment by shahmishal** [comment]
checksum is generated by `swift package compute-checksum` command, its not the same as SHA256.

**Comment by shahmishal** [comment]
The website issue should be resolved now, thanks for letting me know.

**Comment by barnard-b** [comment]
Thank you. The temporary workaround in the 9/17 toolchain and static SDK appears to be working, the test code compiles successfully.

**Comment by briancoyner** [comment]
I can also confirm the 9.17 toolchain + Static Linux SDK appears to resolve the `_FoundationCollections` build issue.

I also wanted to mention that there appears to be another build issue with the Swift 6 toolchain (9.17 version) + SDK issue with SwiftNIO. That issue is reported here: https://github.com/apple/swift-nio/issues/2893

Many thanks!

**Comment by seanmrich** [comment]
I'm not getting the same result. Using the same test code as the OP, I still get the missing `_FoundationCollections` error.

Using Xcode 16 release version
LLVM 43d73cb0cc589f7
Swift aaa632cea622394
Target: arm64-apple-macosx14.0
Linux SDK swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a_static-linux-0.0.1

**Comment by lorentey** [comment]
It looks like this family of issues is now resolved with swift-collections 1.1.4 and the latest swift snapshots. Please reopen if you are still encountering this!
